File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isSpace, isUpper, isLower)
import Data.List (isSuffixOf)

responseFor :: String -> String
responseFor x
  | isYelling && isQuestion = "Calm down, I know what I'm doing!"
  | isQuestion              = "Sure."
  | isYelling               = "Whoa, chill out!"
  | null phrase             = "Fine. Be that way!"
  | otherwise               = "Whatever."
  where
    phrase = filter (not . isSpace) x
    isQuestion = "?" `isSuffixOf` phrase
    isYelling  = any isUpper phrase && (not . any isLower) phrase

File: python/bob/bob.py
def hey(phrase_orig: str) -> str:
    phrase: str = ''.join(phrase_orig.split())
    is_question: bool = phrase.endswith('?')
    is_yelling: bool = phrase.isupper()

    response: str
    if is_yelling and is_question:
        response = "Calm down, I know what I'm doing!"
    elif is_question:
        response = "Sure."
    elif is_yelling:
        response = "Whoa, chill out!"
    elif not phrase:
        response = "Fine. Be that way!"
    else:
        response = "Whatever."

    return response
